File: SimulatorConnectObjCDemo/SimulatorConnectObjCDemo/DeviceTableViewController.m
//
//  DeviceTableViewController.m
//  SimulatorConnectObjCDemo
//
//  Created by Chad Godsey on 5/25/21.
//

#import "DeviceTableViewController.h"
#import <AuthenticationServices/AuthenticationServices.h>
#import "Authentication/AuthenticationPresenter.h"
#import "AppData.h"

@interface DeviceTableViewController () <ASWebAuthenticationPresentationContextProviding, AuthenticationDelegate>

@end

@implementation DeviceTableViewController {
    AuthenticationPresenter *presenter;
    FSGConnect* _connect;
    AppData* _appData;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Get the shared instances for FSGConnect and our app data.
    _connect = FSGConnect.shared;
    _appData = AppData.sharedManager;
}

- (void)viewWillAppear:(BOOL)animated
{
    _appData.devices = [[NSArray alloc] init];
    [[self tableView] reloadData];
    
    // Uncomment the following to show the web login and allow users to sign in
    //presenter = [[AuthenticationPresenter alloc] init];
    //[presenter showSignin: self];
    
    // Uncomment the following to authorize with an app key and start scanning for nearby devices
    NSError __autoreleasing *error = nil;
    [_connect initializeWithAcountId:@"ACCOUNT_ID_HERE" accountKey:@"ACCOUNT_KEY_HERE" error:&error];
    [_connect findDevicesAsyncAndReturnError:&error completion:^(NSArray<id<LMDevice>> * devices, NSError * error) {
        [self findDevicesComplete:devices error:error];
    }];
}

- (void)findDevicesComplete:(NSArray<id<LMDevice>>*)devices error:(NSError*)error {
    _appData.devices = devices;
    [[self tableView] reloadData];
}

#pragma mark - ASWebAuthenticationPresentationContextProviding
- (ASPresentationAnchor)presentationAnchorForWebAuthenticationSession:(ASWebAuthenticationSession *)session {
   return UIApplication.sharedApplication.keyWindow;
}

#pragma mark - AuthenticationDelegate
- (void)authFailure:(NSError *)error
{
    // TODO: Handle error
}

- (void)authSuccess:(NSString *)code
{
    NSError __autoreleasing *error = nil;
    
    // Initialize with user token and scan for devices
    [_connect initializeWithAccessToken:code];
    [_connect findDevicesAsyncAndReturnError:&error completion:^(NSArray<id<LMDevice>> * devices, NSError * error) {
        [self findDevicesComplete:devices error:error];
    }];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _appData.devices.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"deviceCell" forIndexPath:indexPath];
    
    // Configure the cell...
    id<LMDevice> device = [_appData.devices objectAtIndex:indexPath.row];
    cell.textLabel.text = device.name;
    
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return @"Devices";
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    _appData.device = [_appData.devices objectAtIndex:indexPath.row];
}


/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SimulatorConnectSwiftDemo/SimulatorConnectSwiftDemo/DeviceTableViewController.swift
//
//  DeviceTableViewController.swift
//  SimulatorConnectSwiftDemo
//
//  Created by Chad Godsey on 5/20/21.
//

import UIKit
import SimulatorConnect
import AWSMobileClient

class DeviceTableViewController: UITableViewController
{
    private var auth: AuthenticationPresenter!
    private var connect: FSGConnect!
    private var data: AppData!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Get the shared instances for FSGConnect and our app data.
        auth = AuthenticationPresenter.shared
        connect = FSGConnect.shared
        data = AppData.shared
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        data.devices = []
        self.tableView.reloadData()
        
        // Uncomment the following to show the web login and allow users to sign in
//        auth.showSignIn(completion: { code, error in
//            self.data.authCode = code
//            self.connect.initialize(accessToken: code)
//            do {
//                try self.connect.findDevicesAsync(completion: self.findDevicesComplete(devices:error:))
//            } catch {}
//        })
        
        // Uncomment the following to authorize with an app key and start scanning for nearby devices
        do {
            try self.connect.initialize(acountId: "ACCOUNT_ID_HERE", accountKey: "ACCOUNT_KEY_HERE")
            try self.connect.findDevicesAsync(completion: self.findDevicesComplete(devices:error:))
        } catch {}
    }
    
    private func findDevicesComplete(devices: [LMDevice]?, error: Error?)
    {
        guard let devices = devices else { return }
        data.devices = devices
        self.tableView.reloadData()
    }
    
    private func deviceConnection(connected: Bool, error: Error?)
    {
        data.connected = connected
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        // #warning Incomplete implementation, return the number of sections
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return data.devices.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "deviceCell", for: indexPath)

        // Configure the cell...
        let device = data.devices[indexPath.row]
        cell.textLabel?.text = "Device \(device.name) UUID: \(device.id)"

        return cell
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "Devices"
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let device = data.devices[indexPath.row]
        print("Selected device \(device.name)")
        data.device = device
    }

    /*
    // Override to support conditional editing of the table view.
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return true
    }
    */

    /*
    // Override to support editing the table view.
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            // Delete the row from the data source
            tableView.deleteRows(at: [indexPath], with: .fade)
        } else if editingStyle == .insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        }    
    }
    */

    /*
    // Override to support rearranging the table view.
    override func tableView(_ tableView: UITableView, moveRowAt fromIndexPath: IndexPath, to: IndexPath) {

    }
    */

    /*
    // Override to support conditional rearranging of the table view.
    override func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        // Return false if you do not want the item to be re-orderable.
        return true
    }
    */

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
